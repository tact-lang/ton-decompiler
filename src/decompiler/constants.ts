export const RootLayout = {
    instructionsCount: 4,
    instructions: {
        SETCP: 0,
        DICTPUSHCONST: 1,
        DICTIGETJMPZ: 2,
        THROWARG: 3,
    },
} as const
