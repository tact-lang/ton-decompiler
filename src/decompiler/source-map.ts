import {readFileSync} from "node:fs"
import {compileFift} from "../fift/compileFift"

export type ProcedureHash = string

export type SourceMap = Map<ProcedureHash, string>

export async function obtainSourceMap(path: string): Promise<SourceMap> {
    const content = readFileSync(path).toString()
    const result = await compileFift(content, true)
    if (result.status !== "source_map") {
        return new Map()
    }
    return result.sourceMap
}
