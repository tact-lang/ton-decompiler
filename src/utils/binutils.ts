import {BitString} from "@ton/core"
import {Buffer} from "node:buffer"

export const prefixToBin = (prefix: string): BitString => {
    const completionTag = prefix.endsWith("_")
    if (completionTag) {
        prefix = prefix.slice(0, -1)
    }

    const padLength = prefix.length % 2
    const paddedHex = prefix.padEnd(prefix.length + padLength, "0")
    const buffer = Buffer.from(paddedHex, "hex")

    const bits = new BitString(buffer, 0, buffer.byteLength * 8 - padLength * 4)
    if (completionTag) {
        return removeCompletionTag(bits)
    }
    return bits
}

export const removeCompletionTag = (bits: BitString): BitString => {
    const lastSetBitIndex = Array.from({length: bits.length}).findIndex((_, i): boolean =>
        bits.at(bits.length - 1 - i),
    )

    if (lastSetBitIndex === -1) {
        throw new Error("no completion tag")
    }

    return bits.substring(0, bits.length - lastSetBitIndex - 1)
}
