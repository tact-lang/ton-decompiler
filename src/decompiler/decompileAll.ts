// This file is based on code from https://github.com/scaleton-labs/tvm-disassembler

import {Cell, Dictionary, DictionaryValue} from "@ton/core";
import {decompile} from "./decompiler";
import {debugSymbols} from "./knownMethods";
import {AST, BlockNode, InstructionNode, MethodNode, ProcedureNode, ProgramNode, ScalarNode} from "../ast";
import {AssemblerWriter} from "../printer/AssemblerWriter";
import {subcell} from "../utils/subcell";

function decompileCell(args: {
    root: boolean;
    source: Cell;
    offset: { bits: number; refs: number };
    limit: { bits: number; refs: number } | null;
    registerRef?: (ref: Cell) => void;
}): ProgramNode | BlockNode {
    const opcodes = decompile({
        src: args.source,
        offset: args.offset,
        limit: args.limit,
        allowUnknown: false
    });

    // Check if we have a default opcodes of func output
    if (args.root && opcodes.length === 4
        && opcodes[0].op.code === 'SETCP'
        && opcodes[1].op.code === 'DICTPUSHCONST'
        && opcodes[2].op.code === 'DICTIGETJMPZ'
        && opcodes[3].op.code === 'THROWARG') {

        // Load dictionary
        let dictKeyLen = opcodes[1].op.args[0];
        let dictCell = opcodes[1].op.args[1];
        let dict = Dictionary.loadDirect<number, {
            offset: number,
            cell: Cell
        }>(Dictionary.Keys.Int(dictKeyLen), createCodeCell(), dictCell);

        // Extract all methods
        let registeredCells = new Map<string, string>();

        const procedures: ProcedureNode[] = [];

        function extractCallRef(cell: Cell) {
            // Check if we have a call ref
            let callHash = cell.hash().toString('hex');
            if (registeredCells.has(callHash)) {
                return registeredCells.get(callHash)!;
            }

            // Add name to a map and assign name
            let name = '?fun_ref_' + cell.hash().toString('hex').substring(0, 16);
            registeredCells.set(callHash, name);

            const node = decompileCell({
                source: cell,
                offset: {bits: 0, refs: 0},
                limit: null,
                root: false,
                registerRef: extractCallRef,
            });

            procedures.push(
                AST.procedure(
                    callHash,
                    node as BlockNode,
                ),
            );

            return name;
        }

        const methods = [...dict].map(([key, value]): MethodNode => {
            return AST.method(
                key,
                decompileCell({
                    source: value.cell,
                    offset: {bits: value.offset, refs: 0},
                    limit: null,
                    root: false,
                    registerRef: extractCallRef,
                }) as BlockNode,
                value.cell.hash().toString('hex'),
                value.offset,
            )
        });

        return AST.program(methods, procedures);
    }

    // Proceed with a regular decompilation
    const instructions: InstructionNode[] = opcodes.map(op => {
        const opcode = op.op;

        switch (opcode.code) {
            case 'CALLREF':
                if (args.registerRef) {
                    args.registerRef(opcode.args[0]);
                }

                return AST.instruction(
                    'INLINECALLDICT',
                    [AST.reference(opcode.args[0].hash().toString('hex'))],
                    op.offset,
                    op.length,
                    op.hash,
                );

            case 'CALLDICT':
                return AST.instruction(
                    opcode.code,
                    [AST.methodReference(opcode.args[0])],
                    op.offset,
                    op.length,
                    op.hash,
                );

            case 'PUSHCONT':
                return AST.instruction(
                    opcode.code,
                    [
                        decompileCell({
                            source: opcode.args[0],
                            offset: {
                                bits: opcode.args[1],
                                refs: opcode.args[2],
                            },
                            limit: {
                                bits: opcode.args[3],
                                refs: opcode.args[4],
                            },
                            root: false,
                            registerRef: args.registerRef,
                        }) as BlockNode,
                    ],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // Slices
            case 'PUSHSLICE':
            case 'STSLICECONST':
                const slice = subcell({
                    cell: opcode.args[0],
                    offsetBits: opcode.args[1],
                    offsetRefs: opcode.args[2],
                    bits: opcode.args[3],
                    refs: opcode.args[4],
                });

                return AST.instruction(
                    opcode.code,
                    [AST.scalar(slice.toString())],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // Special cases for continuations
            case 'IFREFELSE':
            case 'IFJMPREF':
            case 'IFREF':
            case 'IFNOTREF':
            case 'IFNOTJMPREF':
            case 'IFREFELSEREF':
            case 'IFELSEREF':
            case 'PUSHREFCONT':
                return AST.instruction(
                    opcode.code,
                    [
                        decompileCell({
                            root: false,
                            source: opcode.args[0],
                            offset: {
                                bits: 0,
                                refs: 0,
                            },
                            limit: null,
                            registerRef: args.registerRef,
                        }) as BlockNode,
                    ],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // Globals
            case 'SETGLOB':
            case 'GETGLOB':
                return AST.instruction(
                    opcode.code,
                    [AST.globalVariable(opcode.args[0])],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // Control Registers
            case 'POPCTR':
            case 'PUSHCTR':
                return AST.instruction(
                    opcode.code === 'POPCTR' ? 'POP' : 'PUSH',
                    [AST.controlRegister(opcode.args[0])],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // Stack Primitives
            case 'POP':
            case 'PUSH':
                return AST.instruction(
                    opcode.code,
                    [AST.stackEntry(opcode.args[0])],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // OPCODE s(i) s(j)
            case 'XCHG':
            case 'XCHG2':
            case 'XCPU':
            case 'PUXC':
            case 'PUSH2':
                return AST.instruction(
                    opcode.code,
                    [AST.stackEntry(opcode.args[0]), AST.stackEntry(opcode.args[1])],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // OPCODE s(i) s(j) s(k)
            case 'XCHG3':
            case 'PUSH3':
            case 'XC2PU':
            case 'XCPUXC':
            case 'XCPU2':
            case 'PUXC2':
            case 'PUXCPU':
            case 'PU2XC':
                return AST.instruction(
                    opcode.code,
                    [
                        AST.stackEntry(opcode.args[0]),
                        AST.stackEntry(opcode.args[1]),
                        AST.stackEntry(opcode.args[2]),
                    ],
                    op.offset,
                    op.length,
                    op.hash,
                );

            // All remaining opcodes
            default:
                return AST.instruction(
                    opcode.code as InstructionNode['opcode'],
                    'args' in opcode
                        ? opcode.args.map((arg): ScalarNode => AST.scalar(arg as any))
                        : [],
                    op.offset,
                    op.length,
                    op.hash,
                );
        }
    })

    if (instructions.length === 0) {
        return AST.block(
            [],
            args.source.hash().toString('hex'),
            args.offset.bits,
            0,
        );
    }

    const lastInstruction = instructions[instructions.length - 1];

    return AST.block(
        instructions,
        args.source.hash().toString('hex'),
        args.offset.bits,
        lastInstruction.offset + lastInstruction.length,
    );
}

export function decompileAll(args: { src: Buffer | Cell }) {
    let source: Cell;
    if (Buffer.isBuffer(args.src)) {
        source = Cell.fromBoc(args.src)[0];
    } else {
        source = args.src;
    }

    const ast = decompileCell({
        source,
        offset: {bits: 0, refs: 0},
        limit: null,
        root: true,
    });

    return AssemblerWriter.write(ast, debugSymbols);
}

function createCodeCell(): DictionaryValue<{ offset: number, cell: Cell }> {
    return {
        serialize: (src, builder) => {
            throw Error('Not implemented');
        },
        parse: (src) => {
            let cloned = src.clone(true);
            let offset = src.offsetBits;
            return {offset, cell: cloned.asCell()};
        }
    };
}
