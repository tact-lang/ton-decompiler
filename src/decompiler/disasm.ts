import {Cell, Dictionary, DictionaryValue} from "@ton/core"
import {subslice} from "../utils/subcell"
import {
    DecodedInstruction,
    decodeInstruction,
    DecompiledInstruction,
    NumericValue,
    OperandValue,
    RefValue,
    SliceValue,
} from "./operand-loader"
import {
    BlockNode,
    InstructionArgument,
    InstructionNode,
    MethodNode,
    ProcedureNode,
    ProgramNode,
} from "../ast/ast"
import {createBlock, createInstruction} from "../ast/helpers"
import {RootLayout} from "./constants"
import {getDisplayNumber, hasHint} from "../spec/helpers"
import {LayoutError, OperandError, UnknownOperandTypeError} from "./errors"

export interface DisassembleParams {
    /**
     * The cell to disassemble.
     */
    readonly source: Cell
    /**
     * The offset in the cell to start disassembling from.
     */
    readonly offset?: {bits: number; refs: number}
    /**
     * The limit in the cell to stop disassembling at.
     */
    readonly limit?: {bits: number; refs: number}
}

/**
 * Disassembles a cell into a list of instructions.
 */
export function disassemble(args: DisassembleParams): DecompiledInstruction[] {
    const bitsOffset = args.offset?.bits ?? 0
    const refsOffset = args.offset?.refs ?? 0

    const bitsLimit = args.limit?.bits ?? args.source.bits.length - bitsOffset
    const refsLimit = args.limit?.refs ?? args.source.refs.length - refsOffset

    // Process only a slice of the source cell to support partial disassembly
    const slice = subslice({
        cell: args.source,
        offsetBits: bitsOffset,
        offsetRefs: refsOffset,
        bits: bitsLimit,
        refs: refsLimit,
    })

    const instructions: DecompiledInstruction[] = []
    const hash = args.source.hash().toString("hex")

    while (slice.remainingBits > 0) {
        const opcodeOffset = slice.offsetBits
        const opcode = decodeInstruction(args.source, slice)
        const opcodeLength = slice.offsetBits - opcodeOffset

        instructions.push({
            op: opcode,
            hash,
            offset: opcodeOffset,
            length: opcodeLength,
        })
    }

    // Since every cell can contain references to other cells, we need to disassemble them recursively.
    while (slice.remainingRefs > 0) {
        const source = slice.loadRef()
        instructions.push(...disassemble({source}))
    }

    return instructions
}

export interface DisassembleAndProcessParams extends DisassembleParams {
    readonly onCellReference?: (cell: Cell) => void
}

/**
 * Disassembles a cell into a list of instructions with the help of `disassembly` function
 * and processes them to correctly handle references, calls and operands.
 *
 * This function is a core function of the decompiler.
 */
export function disassembleAndProcess(args: DisassembleAndProcessParams): BlockNode {
    const opcodes = disassemble(args)
    const hash = args.source.hash().toString("hex")
    const offset = args.offset?.bits ?? 0

    const instructions = opcodes.map(op => processInstruction(op, args))

    if (instructions.length === 0) {
        return createBlock([], hash, offset, 0)
    }

    const lastInstruction = instructions.at(-1)
    if (lastInstruction === undefined) throw new Error("unreachable")

    return createBlock(instructions, hash, offset, lastInstruction.offset + lastInstruction.length)
}

/**
 * Processes an instruction to correctly handle references, calls and operands.
 */
function processInstruction(
    op: DecompiledInstruction,
    args: DisassembleAndProcessParams,
): InstructionNode {
    const opcode = op.op
    const opcodeName = opcode.definition.mnemonic

    switch (opcodeName) {
        case "CALLREF": {
            return processCallRef(op, args)
        }
        case "CALLDICT":
        case "CALLDICT_LONG":
        case "JMPDICT": {
            return processCallDict(op)
        }
    }

    return processDefaultInstruction(op, args)
}

/**
 * Processes a `CALLREF` instruction.
 *
 * Decompiler has two modes for decompilation.
 *
 * In bitcode `CALLREF` means that the code of the function
 * placed in the operand of this instruction.
 *
 * In some cases we want to extract the code of the function
 * to another standalone function to make the result more readable.
 */
function processCallRef(
    op: DecompiledInstruction,
    args: DisassembleAndProcessParams,
): InstructionNode {
    const opcode = op.op
    const operand = opcode.operands.find(it => it.definition.name === "c")
    if (!operand || operand.type !== "ref") {
        throw new OperandError(opcode.definition.mnemonic, "c", "Cell", {
            operands: opcode.operands,
        })
    }

    // onCellReference is set when we want to extract the code of the function
    if (args.onCellReference) {
        args.onCellReference(operand.value)
        return createInstruction(
            // actually INLINECALLDICT is Fift opcode, not TVM,
            // but we use it to get same behavior when compiling the resulting
            // Fift asm file back to the TVM code
            PSEUDO_INLINECALLDICT,
            [
                {
                    type: "reference",
                    hash: operand.value.hash().toString("hex"),
                },
            ],
            op.offset,
            op.length,
            op.hash,
        )
    }

    const block = disassembleAndProcess({
        source: operand.value,
        offset: {
            bits: 0,
            refs: 0,
        },
        onCellReference: args.onCellReference,
    })

    return createInstruction(opcode, [{...block, cell: true}], op.offset, op.length, op.hash)
}

/**
 * Processes a `CALLDICT`, `CALLDICT_LONG` or `JMPDICT` instruction.
 */
function processCallDict(op: DecompiledInstruction): InstructionNode {
    const opcode = op.op
    const operand = opcode.operands.find(it => it.definition.name === "n")
    if (!operand || operand.type !== "numeric") {
        throw new OperandError(opcode.definition.mnemonic, "n", "numeric", {
            operands: opcode.operands,
        })
    }

    return createInstruction(
        opcode,
        [
            {
                type: "method_reference",
                methodId: operand.value,
            },
        ],
        op.offset,
        op.length,
        op.hash,
    )
}

/**
 * Processes all other instructions.
 *
 * In other instructions we need to process all its operands if any.
 */
function processDefaultInstruction(
    op: DecompiledInstruction,
    args: DisassembleAndProcessParams,
): InstructionNode {
    const opcode = op.op

    const operands = opcode.operands.map((operand): InstructionArgument => {
        switch (operand.type) {
            case "numeric": {
                return processNumericOperand(operand)
            }
            case "bigint": {
                return {
                    type: "scalar",
                    value: operand.value,
                }
            }
            case "ref":
            case "subslice": {
                return processRefOrSliceOperand(opcode, operand, args)
            }
            default: {
                throw new UnknownOperandTypeError(operand, {
                    instruction: opcode.definition.mnemonic,
                })
            }
        }
    })

    return createInstruction(opcode, operands, op.offset, op.length, op.hash)
}

/**
 * Processes a numeric operand.
 *
 * Numeric operand can be actually be a stack entry, control register or just a scalar.
 */
function processNumericOperand(operand: NumericValue): InstructionArgument {
    const displayHints = operand.definition.display_hints

    // some instructions have hints with value that should be added to the operand
    // for example for `PUSHPOW2` it's 1
    const addHint = displayHints.find(
        (hint): hint is {type: "add"; value: number} => hint.type === "add",
    )
    const add = addHint?.value ?? 0

    const displayNumber = getDisplayNumber(operand, add, displayHints)

    if (hasHint(displayHints, "stack")) {
        return {
            type: "stack_entry",
            value: displayNumber,
        }
    }

    if (hasHint(displayHints, "register")) {
        return {
            type: "control_register",
            value: displayNumber,
        }
    }

    return {
        type: "scalar",
        value: displayNumber,
    }
}

/**
 * Processes a reference or slice operand.
 */
function processRefOrSliceOperand(
    opcode: DecodedInstruction,
    operand: RefValue | SliceValue,
    args: DisassembleAndProcessParams,
): InstructionArgument {
    const displayHints = operand.definition.display_hints
    const opcodeName = opcode.definition.mnemonic

    if (
        hasHint(displayHints, "continuation") ||
        opcodeName === "PUSHCONT" ||
        opcodeName === "PUSHCONT_SHORT"
    ) {
        if (operand.type === "ref") {
            const block = disassembleAndProcess({
                source: operand.value,
                offset: {
                    bits: 0,
                    refs: 0,
                },
                onCellReference: args.onCellReference,
            })
            return {
                ...block,
                cell: true,
            }
        }

        return disassembleAndProcess({
            source: operand.source,
            offset: {
                bits: operand.offsetBits,
                refs: operand.offsetRefs,
            },
            limit: {
                bits: operand.limitBits,
                refs: operand.limitRefs,
            },
            onCellReference: args.onCellReference,
        })
    }

    return {
        type: "scalar",
        value: operand.value.toString(),
    }
}

/**
 * Checks if the root layout is valid.
 *
 * Valid layout is:
 * - `SETCP`
 * - `DICTPUSHCONST`
 * - `DICTIGETJMPZ`
 * - `THROWARG`
 *
 * This is the only layout that is supported by the decompiler.
 * This layout is generated by the FunC and Tact compilers.
 */
function checkLayout(opcodes: DecompiledInstruction[]): void {
    if (opcodes.length !== RootLayout.instructionsCount) {
        throw new LayoutError(RootLayout.instructionsCount, opcodes.length, {
            instructions: opcodes.map(op => op.op.definition.mnemonic),
        })
    }

    const isValidLayout =
        opcodes[RootLayout.instructions.SETCP].op.definition.mnemonic === "SETCP" &&
        opcodes[RootLayout.instructions.DICTPUSHCONST].op.definition.mnemonic === "DICTPUSHCONST" &&
        opcodes[RootLayout.instructions.DICTIGETJMPZ].op.definition.mnemonic === "DICTIGETJMPZ" &&
        opcodes[RootLayout.instructions.THROWARG].op.definition.mnemonic === "THROWARG"

    if (!isValidLayout) {
        throw new LayoutError(RootLayout.instructionsCount, opcodes.length, {
            expected: ["SETCP", "DICTPUSHCONST", "DICTIGETJMPZ", "THROWARG"],
            actual: opcodes.map(op => op.op.definition.mnemonic),
        })
    }
}

/**
 * Disassembles the root cell into a list of instructions.
 *
 * Use this function if you want to disassemble the whole BoC file.
 */
export function disassembleRoot(
    cell: Cell,
    options: {
        /**
         * Whether to deduplicate refs into separate functions. True, by default.
         */
        computeRefs: boolean
    },
): ProgramNode {
    const opcodes = disassemble({source: cell})
    checkLayout(opcodes)

    const dictOpcode = opcodes[RootLayout.instructions.DICTPUSHCONST].op
    const {procedures, methods} = deserializeDict(dictOpcode.operands, options.computeRefs)

    const args = {
        source: cell,
        offset: {bits: 0, refs: 9},
        onCellReference: undefined,
    }
    return {
        type: "program",
        topLevelInstructions: opcodes.map(op => processInstruction(op, args)),
        procedures,
        methods,
        withRefs: options.computeRefs,
    }
}

/**
 * Disassembles a cell without any additional checks for the layout.
 *
 * Use this function if your contract use non-usual layout.
 */
export function disassembleRawRoot(cell: Cell): BlockNode {
    return disassembleAndProcess({
        source: cell,
        onCellReference: undefined,
    })
}

/**
 * Deserializes a dictionary from the dictionary opcode to list
 * of procedures and methods.
 */
function deserializeDict(
    operands: OperandValue[],
    computeRefs: boolean,
): {
    procedures: ProcedureNode[]
    methods: MethodNode[]
} {
    const dictKeyLen = operands.find(operand => operand.definition.name === "n")
    const dictCell = operands.find(operand => operand.definition.name === "d")

    if (!dictKeyLen || !dictCell || dictKeyLen.type !== "numeric" || dictCell.type !== "ref") {
        throw new Error("Cannot find valid operands for Cell and length")
    }

    function createCodeCell(): DictionaryValue<{offset: number; cell: Cell}> {
        return {
            serialize: (_src, _builder): never => {
                throw new Error("Not implemented")
            },
            parse: (src): {offset: number; cell: Cell} => {
                const cloned = src.clone(true)
                const offset = src.offsetBits
                return {offset, cell: cloned.asCell()}
            },
        }
    }

    const countEntries = dictKeyLen.value
    const dict = Dictionary.loadDirect<
        number,
        {
            offset: number
            cell: Cell
        }
    >(Dictionary.Keys.Int(countEntries), createCodeCell(), dictCell.value)

    const registeredCells: Map<string, string> = new Map()

    const procedures: ProcedureNode[] = []

    function extractReferencedCell(cell: Cell): string {
        const callHash = cell.hash().toString("hex")
        const prevCell = registeredCells.get(callHash)
        if (prevCell !== undefined) {
            return prevCell
        }

        const name = "?fun_ref_" + callHash.substring(0, 16)
        registeredCells.set(callHash, name)

        const block = disassembleAndProcess({
            source: cell,
            onCellReference: extractReferencedCell,
        })

        procedures.push({
            type: "procedure",
            hash: callHash,
            body: block,
        })

        return name
    }

    const methods = [...dict].map(([key, value]): MethodNode => {
        return {
            type: "method",
            id: key,
            body: disassembleAndProcess({
                source: value.cell,
                offset: {bits: value.offset, refs: 0},
                onCellReference: computeRefs ? extractReferencedCell : undefined,
            }),
            hash: value.cell.hash().toString("hex"),
            offset: value.offset,
        }
    })

    return {
        procedures,
        methods,
    }
}

const PSEUDO_INLINECALLDICT = {
    definition: {
        mnemonic: "INLINECALLDICT",
        doc: {
            fift: "",
            opcode: "",
            gas: "",
            category: "",
            stack: "",
            description: "",
            fift_examples: [],
        },
        bytecode: {
            operands: [],
            tlb: "",
            prefix: "",
        },
        control_flow: {
            branches: [],
            nobranch: true,
        },
        value_flow: {
            inputs: {
                stack: [],
                registers: [],
            },
            outputs: {
                stack: [],
                registers: [],
            },
        },
        since_version: 0,
    },
    operands: [],
}
