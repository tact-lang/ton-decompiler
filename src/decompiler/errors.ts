import {OperandValue} from "./operand-loader"

export class DisassemblerError extends Error {
    public constructor(message: string, details?: Record<string, unknown>) {
        const detailsStr = details
            ? "\nDetails:\n" +
              Object.entries(details)
                  .map(([key, value]) => `  ${key}: ${JSON.stringify(value, null, 2)}`)
                  .join("\n")
            : ""

        super(message + detailsStr)
        this.name = "DisassemblerError"
    }
}

export class OperandError extends DisassemblerError {
    public constructor(
        opcode: string,
        operandName: string,
        expectedType: string,
        details?: Record<string, unknown>,
    ) {
        super(`Cannot find ${expectedType} operand "${operandName}" for "${opcode}"`, {
            opcode,
            operandName,
            expectedType,
            ...details,
        })
        this.name = "OperandError"
    }
}

export class LayoutError extends DisassemblerError {
    public constructor(expected: number, actual: number, details?: Record<string, unknown>) {
        super(`Unexpected root layout: expected ${expected} instructions, got ${actual}`, {
            expected,
            actual,
            ...details,
        })
        this.name = "LayoutError"
    }
}

export class UnknownOperandTypeError extends DisassemblerError {
    public constructor(operand: OperandValue, details?: Record<string, unknown>) {
        super(`Unknown operand type: ${operand.type}`, {
            type: operand.type,
            ...details,
        })
        this.name = "UnknownOperandTypeError"
    }
}
