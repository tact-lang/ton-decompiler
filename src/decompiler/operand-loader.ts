import {BitString, Builder, Cell, Slice} from "@ton/core"
import type {Instruction, Operand} from "../spec/tvm-spec"
import {removeCompletionTag} from "../utils/binutils"
import {PrefixMatcher} from "../utils/prefix-matcher"
import {DisassemblerError} from "./errors"
import {repeat} from "../utils/tricks"

type ExtractType<T extends Operand["type"], A = Operand> = A extends {type: T} ? A : never

export type UintOperand = ExtractType<"uint">
export type IntOperand = ExtractType<"int">
export type PushLongOperand = ExtractType<"pushint_long">
export type SubsliceOperand = ExtractType<"subslice">
export type RefOperand = ExtractType<"ref">

export type NumericOperand = UintOperand | IntOperand

export interface NumericValue {
    readonly type: "numeric"
    readonly definition: NumericOperand
    readonly value: number
    readonly bitcode: BitString
}

export interface BigIntValue {
    readonly type: "bigint"
    readonly definition: PushLongOperand
    readonly value: bigint
    readonly bitcode: BitString
}

export interface RefValue {
    readonly type: "ref"
    readonly definition: RefOperand
    readonly value: Cell
    readonly bitcode: BitString
}

export interface SliceValue {
    readonly type: "subslice"
    readonly definition: SubsliceOperand
    readonly value: Cell
    readonly source: Cell
    readonly offsetBits: number
    readonly offsetRefs: number
    readonly limitBits: number
    readonly limitRefs: number
}

export type OperandValue = NumericValue | BigIntValue | RefValue | SliceValue

export interface DecodedInstruction {
    readonly definition: Instruction
    readonly operands: OperandValue[]
}

export interface DecompiledInstruction {
    readonly op: DecodedInstruction
    readonly hash: string
    readonly offset: number
    readonly length: number
}

const prefixMatcher: PrefixMatcher = new PrefixMatcher()

export function decodeInstruction(source: Cell, slice: Slice): DecodedInstruction {
    const definition = prefixMatcher.loadPrefix(slice)
    const operands = parseInstructions(source, slice, definition)
    return {
        definition,
        operands,
    }
}

export function parseInstructions(
    source: Cell,
    slice: Slice,
    instruction: Instruction,
): OperandValue[] {
    const operands: OperandValue[] = []
    for (const operand of instruction.bytecode.operands) {
        try {
            operands.push(parseInstruction(source, operand, slice))
        } catch (error) {
            throw new DisassemblerError(
                `Bad operand ${operand.name} for instruction ${instruction.mnemonic}`,
                {
                    cause: error,
                },
            )
        }
    }
    return operands
}

function parseInstruction(source: Cell, operand: Operand, slice: Slice): OperandValue {
    const type = operand.type

    if (type === "uint") {
        const raw = slice.clone().loadBits(operand.size)
        return {
            type: "numeric",
            definition: operand,
            bitcode: raw,
            value: slice.loadUint(operand.size),
        }
    }

    if (type === "int") {
        const raw = slice.clone().loadBits(operand.size)
        return {
            type: "numeric",
            definition: operand,
            bitcode: raw,
            value: slice.loadInt(operand.size),
        }
    }

    if (type === "ref") {
        const raw = slice.clone().loadRef()
        return {type: "ref", definition: operand, bitcode: raw.bits, value: slice.loadRef()}
    }

    if (type === "pushint_long") {
        const cloned = slice.clone()

        const prefix = slice.loadUint(5)
        const length = 8 * prefix + 19

        const raw = cloned.loadBits(5 + length)
        return {
            type: "bigint",
            definition: operand,
            bitcode: raw,
            value: slice.loadIntBig(length),
        }
    }

    // Handle remaining subslice type

    const refLength =
        (operand.refs_add ?? 0) +
        (operand.refs_length_var_size ? slice.loadUint(operand.refs_length_var_size) : 0)
    const bitLength =
        operand.bits_padding +
        (operand.bits_length_var_size ? slice.loadUint(operand.bits_length_var_size) * 8 : 0)

    const offsetBits = slice.offsetBits
    const offsetRefs = slice.offsetRefs

    const loadedBits = slice.loadBits(bitLength)
    const bits = operand.completion_tag ? removeCompletionTag(loadedBits) : loadedBits

    const builder = new Builder()
    builder.storeBits(bits)

    repeat(refLength, () => builder.storeRef(slice.loadRef()))

    return {
        type: "subslice",
        definition: operand,
        value: builder.endCell(),
        source: source,
        offsetBits,
        offsetRefs,
        limitBits: bitLength,
        limitRefs: refLength,
    }
}
