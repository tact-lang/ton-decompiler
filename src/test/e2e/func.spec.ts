import {describe, it, expect} from "vitest"
import {Cell} from "@ton/core"
import {disassembleRoot} from "../../decompiler/disasm"
import {AssemblyWriter} from "../../printer/assembly-writer"
import {debugSymbols} from "../../utils/known-methods"
import {compileFunc} from "@ton-community/func-js"
import {fail} from "node:assert"
import {compileFiftBackAndCompare} from "./utils"

describe("disassemble", () => {
    it("should decompile and compile back PUSHREF with bytes", async () => {
        const funcRes = await compileFunc({
            sources: [
                {
                    filename: "main.fc",
                    content: `
                    int cell_hash(cell c) asm "HASHCU";

                    cell __gen_cell_cell_37e90db9d1f7725dc0128ee6bad2035fb50479e09a488a29257bed01a23050a0() asm """
                        B{b5ee9c7241010101000d00001600000000537563636573738a3a2a2a} B>boc PUSHREF
                    """;

                    () recv_internal() {
                        cell c = __gen_cell_cell_37e90db9d1f7725dc0128ee6bad2035fb50479e09a488a29257bed01a23050a0();
                        throw(cell_hash(c));
                    }`,
                },
            ],
        })

        if (funcRes.status === "error") {
            fail(`cannot compile FunC: ${funcRes.message}`)
            return
        }

        const buffer = Buffer.from(funcRes.codeBoc, "base64")

        const res = disassembleRoot(Cell.fromBoc(buffer)[0], {
            computeRefs: false,
        })

        const decompiled = AssemblyWriter.write(res, {
            useAliases: true,
            withoutHeader: true,
            debugSymbols: debugSymbols,
        })

        expect(decompiled).toMatchSnapshot()

        await compileFiftBackAndCompare(decompiled, buffer)
    })
})
