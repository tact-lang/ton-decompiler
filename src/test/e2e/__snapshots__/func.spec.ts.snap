// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`disassemble > should decompile GETGLOB and GETGLOBVAR correctly 1`] = `
""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  recv_internal PROC:<{
    ONE
    ONE
    ADD
    GETGLOBVAR
    DROP
    2 GETGLOB
    DROP
  }>
}END>c"
`;

exports[`disassemble > should decompile SDBEGINSXQ correctly 1`] = `
""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  recv_internal PROC:<{
    NEWC
    x{68656C6C6F20776F726C64} PUSHSLICE
    STSLICER
    ENDC
    CTOS
    NEWC
    x{68656C6C6F} PUSHSLICE
    STSLICER
    ENDC
    CTOS
    DUP2
    SDBEGINSX
    ROTREV
    SDBEGINSXQ
    DROP2
  }>
}END>c"
`;

exports[`disassemble > should decompile and compile back PUSHREF with bytes 1`] = `
""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  recv_internal PROC:<{
    <b x{0000000053756363657373} s, b> PUSHREF
    HASHCU
    THROWANY
  }>
}END>c"
`;
